cgo/compare.c

```diff
@@ -0,0 +1,370 @@
+    } else {                                                  \
+        if (nulls != NULL) {                                  \
+            for (uint64_t i = 0; i < n; i++) {                \
+                if (!bitmap_test(nulls, i)) {                 \
+                    OP(rt[i], at[i], bt[i]);                  \
+                }                                             \
+            }                                                 \
+        } else {                                              \
+            for (uint64_t i = 0; i < n; i++) {                \
+                OP(rt[i], at[i], bt[i]);                      \
+            }                                                 \
+        }                                                     \
+    }
+
+
+
+int32_t Numeric_VecEq(void *r, void *a, void  *b, uint64_t n, uint64_t *nulls, int32_t flag, int32_t type)
+{
+    if (type == Type_INT8) {
+        MO_COMPARE_T(COMPARE_EQ, int8_t);
+    } else if (type == Type_INT16) {
+        MO_COMPARE_T(COMPARE_EQ, int16_t);
+    } else if (type == Type_INT32) {
+        MO_COMPARE_T(COMPARE_EQ, int32_t);
+    } else if (type == Type_INT64) {
+        MO_COMPARE_T(COMPARE_EQ, int64_t);
+    } else if (type == Type_UINT8) {
+        MO_COMPARE_T(COMPARE_EQ, uint8_t);
+    } else if (type == Type_UINT16) {
+        MO_COMPARE_T(COMPARE_EQ, uint16_t);
+    } else if (type == Type_UINT32) {
+        MO_COMPARE_T(COMPARE_EQ, uint32_t);
+    } else if (type == Type_UINT64) {
+        MO_COMPARE_T(COMPARE_EQ, uint64_t);
+    } else if (type == Type_FLOAT32) {
+        MO_COMPARE_T(COMPARE_EQ, float);
+    } else if (type == Type_FLOAT64) {
+        MO_COMPARE_T(COMPARE_EQ, double);
+    } else if (type == Type_DATE) {
+        MO_COMPARE_T(COMPARE_EQ, int32_t);
+    } else if (type == Type_DATETIME) {
+        MO_COMPARE_T(COMPARE_EQ, int64_t);
+    } else if (type == Type_TIMESTAMP) {
+        MO_COMPARE_T(COMPARE_EQ, int64_t);
+    } else if (type == Type_BOOL) {
+        MO_COMPARE_T(COMPARE_BOOL_EQ, bool);
+    } else {
+        return RC_INVALID_ARGUMENT;
+    }
+    return RC_SUCCESS;
+}
+
+
+int32_t Numeric_VecNe(void *r, void *a, void  *b, uint64_t n, uint64_t *nulls, int32_t flag, int32_t type)
+{
+    if (type == Type_INT8) {
+        MO_COMPARE_T(COMPARE_NE, int8_t);
+    } else if (type == Type_INT16) {
+        MO_COMPARE_T(COMPARE_NE, int16_t);
+    } else if (type == Type_INT32) {
+        MO_COMPARE_T(COMPARE_NE, int32_t);
+    } else if (type == Type_INT64) {
+        MO_COMPARE_T(COMPARE_NE, int64_t);
+    } else if (type == Type_UINT8) {
+        MO_COMPARE_T(COMPARE_NE, uint8_t);
+    } else if (type == Type_UINT16) {
+        MO_COMPARE_T(COMPARE_NE, uint16_t);
+    } else if (type == Type_UINT32) {
+        MO_COMPARE_T(COMPARE_NE, uint32_t);
+    } else if (type == Type_UINT64) {
+        MO_COMPARE_T(COMPARE_NE, uint64_t);
+    } else if (type == Type_FLOAT32) {
+        MO_COMPARE_T(COMPARE_NE, float);
```

Comment:
How to handle the approximate equal on the float and double type?

Reply:
Float and double only compare the same types, and can be compared directly. so there is no case of  approximate equal.